### PR TITLE
Modify patterns from a skeleton match to have the correct width

### LINKED ISSUE #584: Modify patterns from a skeleton match to have the correct width
https://unicode.org/reports/tr35/tr35-dates.html#Matching_Skeletons

>  * 2.6.2.1 Matching Skeletons
>    - TODO - Modify the resulting pattern to have fields of the same length. For example requesting
>      a skeleton "yMMMMd" can have a best match of ["yMMMd", "d MMM y"]. This pattern should
>      then be modified to use the requested length to produce a pattern "d MMMM y".
>      However, fields should not be changed from numeric to text.

This is a follow-up to #481. It blocks #272.

### PR DESCRIPTION
**Pull request by gregtatum**
Resolves #584.

### CODE REVIEW

File: components/datetime/src/skeleton.rs
Comment (zbraniecki) [inline components/datetime/src/skeleton.rs:625]: `suggestion (n-b)`: the only uses of `get_length_type` is to retrieve it for two fields and compare if they're equal. Could we optimize it a bit by having `field.matches_length(&self, field) -> bool`?

**Review by zbraniecki — APPROVED** [review]

File: components/datetime/src/skeleton.rs
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:625]: I'm not sure what you mean by "optimize", but I would prefer the current implementation to:

```diff
diff --git a/components/datetime/src/fields/mod.rs b/components/datetime/src/fields/mod.rs
index 30652acc..829f8fc8 100644
--- a/components/datetime/src/fields/mod.rs
+++ b/components/datetime/src/fields/mod.rs
@@ -31,7 +31,11 @@ pub struct Field {
 }
 
 impl Field {
-    pub fn get_length_type(&self) -> TextOrNumeric {
+    pub fn matches_length_type(&self, field: &Field) -> bool {
+        self.get_length_type() == field.get_length_type()
+    }
+
+    fn get_length_type(&self) -> TextOrNumeric {
         match self.symbol {
             FieldSymbol::Year(year) => year.get_length_type(self.length),
             FieldSymbol::Month(month) => month.get_length_type(self.length),
diff --git a/components/datetime/src/skeleton.rs b/components/datetime/src/skeleton.rs
index 7f5836a6..dd4f358a 100644
--- a/components/datetime/src/skeleton.rs
+++ b/components/datetime/src/skeleton.rs
@@ -565,11 +565,10 @@ pub fn get_best_available_format_pattern(
                         NO_DISTANCE
                     } else if requested_field.symbol != skeleton_field.symbol {
                         SUBSTANTIAL_DIFFERENCES_DISTANCE
-                    } else if requested_field.get_length_type() != skeleton_field.get_length_type()
-                    {
-                        TEXT_VS_NUMERIC_DISTANCE
-                    } else {
+                    } else if requested_field.matches_length_type(&skeleton_field) {
                         WIDTH_MISMATCH_DISTANCE
+                    } else {
+                        TEXT_VS_NUMERIC_DISTANCE
                     };
 
                     requested_fields.next();
@@ -622,7 +621,7 @@ pub fn get_best_available_format_pattern(
                         .find(|field| field.symbol == pattern_field.symbol)
                     {
                         if requested_field.length != pattern_field.length
-                            && requested_field.get_length_type() == pattern_field.get_length_type()
+                            && requested_field.matches_length_type(&pattern_field)
                         {
                             return PatternItem::Field(*requested_field);
                         }
```

Comment (zbraniecki) [reply components/datetime/src/skeleton.rs:625]: yup, that's what I suggesed. If you prefer the current API, I'm okay with that!

File: components/datetime/src/skeleton.rs
Comment (sffc) [inline components/datetime/src/skeleton.rs:634]: Thought: I wish we could do this without `alloc`.  I liked how before this function was basically a projection from `&'a SkeletonsV1` to `&'a Pattern`.  The extra allocation is going to hurt basically every benchmark we're after: memory usage, code size, and performance.

Suggestion: Could this be architected in a way that involves a map on the pattern itself?  Like, you could attach a mapping function to the pattern, such that when you loop over the pattern while formatting, you run the PatternItems through the mapping function.

Alternatively, is this in the part of the code that we won't use at runtime after the new CLDR 70 skeleton algorithm is implemented?  I don't care about the `Vec` if it's run in the transformer.

**Review by sffc — DISMISSED** [review]
Not really requesting changes, but I have a question before I approve

File: components/datetime/src/skeleton.rs
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:634]: > The extra allocation is going to hurt basically every benchmark we're after.

Yes, I agree that the allocation isn't great, and what I tried to design around in the initial implementation.

> Suggestion: Could this be architected in a way that involves a map on the pattern itself.

The pattern _must_ be mutated in order to support all of the features for DateTimeFormat. Beyond just the widths, there's also the hour cycle to consider. There's also append items support for things like time zones or week. The latter is even trickier, as a simple mapping function wouldn't work.

I would prefer to accept the perf/memory regression for 0.3, and follow-up with exploring some approaches. Especially as pattern mutation is a new requirement for other features such as the hour cycle, and append items. Then for 0.4 I can track fixing it.

> Alternatively, is this in the part of the code that we won't use at runtime after the new CLDR 70 skeleton algorithm is implemented? I don't care about the Vec if it's run in the transformer.

I don't think well want a literal pattern for every width adjustment, as it will greatly affect the provider data. I'd say let's follow-up with looking for perf wins here. I'm nervous about the trade off between runtime characteristics and data payload size. I think it's something we should definitely try and work out.

Comment (gregtatum) [reply components/datetime/src/skeleton.rs:634]: I filed https://github.com/unicode-org/icu4x/issues/877 as a follow-up, if you are OK with moving forward with this patch.

Comment (sffc) [reply components/datetime/src/skeleton.rs:634]: Okay.  I won't block this PR but I'm a bit disappointed that we're going in this direction.  I feel like we're repeating the problems of ICU4C rather than solving them.  I hope we can follow up in #877.

Comment (gregtatum) [reply components/datetime/src/skeleton.rs:634]: Well, this at least gets us a baseline metric that we can track, and tests in place that make sure we are compliant with the behavior. The code is mutable, so we can course correct to fewer allocations. I also filed #879 to make sure we track this area better. I realize now that the benchmark is still needed.

**Pull request merged**
